engine/common/follower/pending_tree/pending_tree.go

```diff
@@ -0,0 +1,198 @@
+package pending_tree
+
+import (
+	"fmt"
+
+	"golang.org/x/exp/slices"
+
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/forest"
+)
+
+// CertifiedBlock holds a certified block, it consists of a block and a QC which proves validity of block (QC.BlockID = Block.ID())
+// This is used to compactly store and transport block and certifying QC in one structure.
+type CertifiedBlock struct {
+	Block *flow.Block
+	QC    *flow.QuorumCertificate
+}
+
+// ID returns unique identifier for the certified block
+// To avoid computation we use value from the QC
+func (b *CertifiedBlock) ID() flow.Identifier {
+	return b.QC.BlockID
+}
+
+// View returns view where the block was produced.
+func (b *CertifiedBlock) View() uint64 {
+	return b.QC.View
+}
+
+// Height returns height of the block.
+func (b *CertifiedBlock) Height() uint64 {
+	return b.Block.Header.Height
+}
+
+// PendingBlockVertex wraps a block proposal to implement forest.Vertex
+// so the proposal can be stored in forest.LevelledForest
+type PendingBlockVertex struct {
+	CertifiedBlock
+	connectedToFinalized bool
```

Comment:
As we have discussed I think that `orphaned` is relatively useless since it can't grow indefinitely, including blocks in orphaned fork should be ok. This leaves is with two states, `connected` and `!connected == disconnected`

Reply:
👍 Agee 💯 If it adds notable complexity to the implementation, it is not worth it.